imageio/core/imopen.py

```diff
@@ -0,0 +1,254 @@
+from numbers import Number
+import re
+import numpy as np
+
+from .format import FormatManager, MODENAMES
+from .request import Request, RETURN_BYTES
+from .util import Singleton
+
+
+class _imopen(metaclass=Singleton):
```

Comment:
I am not sure we need to use the `Singleton` metaclass here... If `_known_plugins` is a class attribute instead of an instance attribute, modules are not constantly re-registred...

Then no need for the definition of `_imopen` and moreover `imopen` can be called directly (`imopen(*args, **kwargs)`).
The `__enter__` and `__exit__` methods can also be directly implemented in `imopen`...

Reply:
We can use a `_known_plugins` class attribute if we decide that we don't want anything to inherit from `imopen`. Especially for complex types like `list` or `dict` this can lead to hard to find bugs and unexpected behavior because we are now sharing state between instances of parent classes and child classes.

```
class A(object):
    _known_plugins = ["A", "B", "C"]

class B(A):
    def __init__(self):
        self._known_plugins.append("D")

b = B()
a = A()
print(f"A knows {a._known_plugins}", f"B knows {b._known_plugins}")
#  prints: A knows ['A', 'B', 'C', 'D'] B knows ['A', 'B', 'C', 'D']
```

This means that `b` is now sharing state with `a`. I don't know if this is a good idea or not. You run the risk of downstream code relying on `a` having a "D" plugin, which suddenly disappears when code in a sister module gets refactored for example like

```
class A(object):
    _known_plugins = ["A", "B", "C"]

class B(A):
    def __init__(self):
        # only use D to avoid conflict with 
        # latest features of plugin A
        self._known_plugins = ["D"]

b = B()
a = A()
print(f"A knows {a._known_plugins}", f"B knows {b._known_plugins}")
# prints: A knows ['A', 'B', 'C'] B knows ['D']
```

Now all code that relied on instances of `A` knowing plugin D will break with a cryptic error message like `"No plugin to load images of type XYZ"`, which doesn't indicate at all that changes to `B` are the cause. Also, `A` and `B` no longer share state, but a new class `C` will still share state with `A` when it naively inherits from `B`.

```
class C(B):
    def __init__(self):
        # plugins are shared with parent class
        self._known_plugins.append("E")

print(C()._known_plugins)
# prints ['A', 'B', 'C', 'E']
```

but 

```
class C(B):
    def __init__(self):
        # maybe not so shared with parent class
        super(C, self).__init__()
        self._known_plugins.append("E")

print(f"C knows {C()._known_plugins}", f"B knows {B()._known_plugins}")
# prints: C knows ['D', 'E'] B knows ['D']
```

I'd rather have it clean and at all times say "if you want to use parent's plugins, call the parent's constructor via `super`".

---

We will still have to use `_imopen` trickery. This comes from `imopen` being a factory, not from it being a singleton. `imopen()` will still be a constructor call to `imopen.__init__` and return an instance of `imopen`, where we actually want it to call `imopen.__call__` and return an instance of a plugin class. Maybe I'm missing something though x)

---

I think it is cleaner to have `__enter__` and `__exit__` methods defined in the plugin. Not every plugin will want to keep a file open for a prolonged period of time and some might want to keep multiple files open, e.g., if you implement a `DatasetFeeder` plugin that takes a `CSV` file with locations of images, stacks them up into batches and feeds them to the user, perhaps including some repetition.

The latter is a more esoteric example (will become more relevant once we move towards reading video though), but the former is true for all legacy plugins.

Reply:
I'm not sure I follow this ... One of the ideas of the new plugin system is that plugins dont need to inherit from anything.

Reply:
No, plugins don't inherit anything. They will have to comply with a minimal API (accept `uri` in the constructor, implement `read(...)`, `write(...)`, etc.), but outside of that, they are pretty free.

What I am talking about is downstream code deciding to inherit the `imopen` class. Not sure why somebody would want to do this, but if a class is part of the public API, one would assume that it is okay to inherit from it unless otherwise specified. Bottom line: We can either explicitly document to not inherit from `imopen` and use class variables (as @rfezzani suggested), or we should stick with a metaclass, which takes inheritance into account.

Reply:
Can it not simply be *not* part of the public API?

Reply:
I removed the `Singleton`, which should resolve this discussion in the way you @almarklein and @rfezzani suggested.